plugin/lighthouse/handler.go

```diff
@@ -0,0 +1,61 @@
+package lighthouse
+
+import (
+	"context"
+	"net"
+	"strings"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/nonwriter"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+// ServeDNS implements the plugin.Handler interface.
+func (lh *Lighthouse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+
+	a := new(dns.Msg)
+	a.SetReply(r)
+	a.Authoritative = true
+
+	services := lh.SvcsMap
+	svcName := strings.Split(state.QName(), ".")[0]
+	ip := services[svcName]
+	if ip == "" {
+		return onFailure(lh, ctx, w, r)
+	}
+	rr := new(dns.A)
+
+	switch state.Family() {
+	case 1:
+		rr.Hdr = dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeA, Class: state.QClass()}
+		rr.A = net.ParseIP(ip).To4()
+	default:
+		log.Debugf("IPv6 queries not supported yet")
+		return onFailure(lh, ctx, w, r)
+	}
+
+	a.Answer = []dns.RR{rr}
+
+	log.Debugf("Responding to query with '%s'", a.Answer)
+	if w.WriteMsg(a) != nil {
+		log.Debugf("Failed to writeMsg")
+		return onFailure(lh, ctx, w, r)
```

Comment:
In this case we're actually handling the query but failed to write the message. On success, we return and of course don't invoke the next plugin so should we defer to the next plugin on failure? It seems we should return a failure code along with the error here.

Reply:
See above. User will know of error as NXDOMAIN response. WriteMsg failing is like basic io failure. Other plugins don't even check for return value and assume success. This was added just to satisfy GoLint.

Reply:
The full sequence in this scenario is that we fail to write the message, try to pass the query on to the next plugin, and if that fails to find anything, the whole process will end up returning `NXDOMAIN`. We’re not indicating anything as a result of the error, and that bothers me — imagine you’re trying to debug this, as an admin; if you see “no such domain” (or some such error, “no such service” or whatever), will you suspect that there *is* really a matching service but that a technical error prevented the appropriate CoreDNS plugin from finding it?

Reply:
Got it, agree it will help troubleshoot if we do run into this error. And if it is that rare, warning/error shouldn't really show up at all. Will add a warn.